### PR TITLE
Portable selected op workaround for arm backend

### PR DESCRIPTION
**Pull request by kirklandsign**
Arm uses portable yaml, and it only needs softmax. However, now we disabled root op list, and can only support yaml, so this breaks arm build.

We temporarily support yaml and ops list while we make a selective build rule for arm.

### CODE REVIEW

**Comment by larryliu0820** [comment]
Can you take the chance and implement it properly? We shouldn't depend on `portable_lib` as a whole, instead we should use selective build to build our own library for ARM. Also it seems we don't have CI for arm?

**Review by larryliu0820 — CHANGES_REQUESTED** [review]
Can we not add selective build feature in `portable/CMakeLists.txt`? Now is a good time to cleanup arm `CMakeLists.txt` by adding a new executable and use selective build for it.

File: examples/arm/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+# Let files say "include <executorch/path/to/header.h>".
+set(_common_include_directories ${EXECUTORCH_ROOT}/..)
+
+include(${EXECUTORCH_ROOT}/build/Utils.cmake)
+include(${EXECUTORCH_ROOT}/build/Codegen.cmake)
+
+# Generate C++ bindings to register kernels into both PyTorch (for AOT) and
+# Executorch (for runtime). Here select all ops in functions.yaml
+gen_selected_ops("" "${EXECUTORCH_SELECT_OPS_LIST}" "")
```
Comment (larryliu0820) [inline examples/arm/CMakeLists.txt:42]: You can probably hardcode the list to be "aten::softmax.out"

File: kernels/portable/CMakeLists.txt [hunk 2]
```diff
@@ -39,8 +39,11 @@ list(FILTER _portable_kernels__srcs EXCLUDE REGEX "test/*.cpp")
 list(FILTER _portable_kernels__srcs EXCLUDE REGEX "codegen")
 # Generate C++ bindings to register kernels into both PyTorch (for AOT) and
 # Executorch (for runtime). Here select all ops in functions.yaml
-set(_yaml "${CMAKE_CURRENT_LIST_DIR}/functions.yaml")
-gen_selected_ops("${_yaml}" "" "")
+if(EXECUTORCH_SELECT_OPS_YAML)
+  set(_yaml "${CMAKE_CURRENT_LIST_DIR}/functions.yaml")
+endif()
+gen_selected_ops(
+  "${_yaml}" "${EXECUTORCH_SELECT_OPS_LIST}" "${EXECUTORCH_SELECT_ALL_OPS}")
```
Comment (larryliu0820) [inline kernels/portable/CMakeLists.txt:46]: remove this change?

File: examples/arm/CMakeLists.txt (hunk 1 above)
Comment (kirklandsign) [reply examples/arm/CMakeLists.txt:45]: I used -DEXECUTORCH_SELECT_OPS_LIST="aten::_softmax.out"  instead in case we want to use something else.

Comment (larryliu0820) [reply examples/arm/CMakeLists.txt:45]: Can you move this option from root CMakeLists.txt to this file? I think this is the only usage.

File: examples/arm/run.sh [hunk 3]
```diff
@@ -59,6 +59,33 @@ function build_executorch() {
     mkdir -p "${et_build_dir}"
 
     cd "${et_build_dir}"
+    cmake                                                 \
+        -DBUCK2=${buck2}                                  \
+        -DCMAKE_INSTALL_PREFIX=cmake-out \
```
Comment (digantdesai) [inline examples/arm/run.sh:64]: nit align `\` :P

File: examples/arm/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+# ~~~
+# cmake-format --first-comment-is-literal=True CMakeLists.txt
+# ~~~
+
+cmake_minimum_required(VERSION 3.19)
+project(arm_example)
+
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
+if(NOT CMAKE_CXX_STANDARD)
+  set(CMAKE_CXX_STANDARD 17)
```
Comment (digantdesai) [inline examples/arm/CMakeLists.txt:17]: curious why 17?

File: examples/arm/CMakeLists.txt (hunk 1 above)
Comment (digantdesai) [reply examples/arm/CMakeLists.txt:45]: yeah Ok to move it here, but we do want it configurable on cmdline

File: examples/arm/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,47 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (digantdesai) [inline examples/arm/CMakeLists.txt:1]: this feels like a boilerplate code, are we sure we want everyone who wants to use selective build write this? That was my thought when I added a list variable at the top level so that some random build can also just "use it". If the answer is yes, I am ok with it too.

File: CMakeLists.txt [hunk 6]
```diff
@@ -340,13 +340,6 @@ if(EXECUTORCH_BUILD_QNN)
   add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/examples/qualcomm)
 endif()
 
-# Build Arm Baremetal backend
-option(EXECUTORCH_BUILD_ARM_BAREMETAL
```
Comment (digantdesai) [inline CMakeLists.txt:344]: we want this switch and a way to build backend from the top level for the top level executor_runner target because examples is optional i.e. end-users may not need that, but this is to gate building the arm backend and should be same for all backends.

File: examples/arm/run.sh [hunk 7]
```diff
@@ -69,15 +96,14 @@ function build_executorch() {
         -DEXECUTORCH_BUILD_ARM_BAREMETAL=ON               \
         -DCMAKE_BUILD_TYPE=Release                        \
         -DEXECUTORCH_ENABLE_LOGGING=ON                    \
-        -DEXECUTORCH_SELECT_OPS_LIST="aten::_softmax.out" \
         -DFLATC_EXECUTABLE="$(which flatc)"               \
         -DCMAKE_TOOLCHAIN_FILE="${toolchain_cmake}"       \
-        "${et_root_dir}"
+        -DEXECUTORCH_SELECT_OPS_LIST="aten::_softmax.out" \
```
Comment (digantdesai) [inline examples/arm/run.sh:101]: why move it down?

File: examples/arm/run.sh [hunk 8]
```diff
@@ -28,7 +28,7 @@ setup_path_script=${root_dir}/setup_path.sh
 
 # Executorch
 et_root_dir=$(cd ${script_dir}/../.. && pwd)
-et_build_dir=${root_dir}/executorch-cmake-out
+et_build_dir=${root_dir}/cmake-out
```
Comment (digantdesai) [inline examples/arm/run.sh:31]: what's the rationale behind this change? can't we install in the `${et_build_dir}`?

File: examples/arm/CMakeLists.txt (hunk 1 above)
Comment (kirklandsign) [reply examples/arm/CMakeLists.txt:45]: Also selective_build/test_selective_build.sh?

File: examples/arm/CMakeLists.txt (hunk 4 above)
Comment (kirklandsign) [reply examples/arm/CMakeLists.txt:17]: Probably doesn't matter. Removed it.

File: examples/arm/CMakeLists.txt (hunk 5 above)
Comment (kirklandsign) [reply examples/arm/CMakeLists.txt:1]: I'm ok with users adding stuff like this.  It's a clean way for them to know what they are doing.

File: CMakeLists.txt (hunk 6 above)
Comment (kirklandsign) [reply CMakeLists.txt:344]: @larryliu0820

File: examples/arm/run.sh [hunk 9]
```diff
@@ -72,12 +100,10 @@ function build_executorch() {
         -DEXECUTORCH_SELECT_OPS_LIST="aten::_softmax.out" \
         -DFLATC_EXECUTABLE="$(which flatc)"               \
         -DCMAKE_TOOLCHAIN_FILE="${toolchain_cmake}"       \
-        "${et_root_dir}"
+        -B${et_build_dir}/examples/arm \
+        "${et_root_dir}"/examples/arm --debug-find
```
Comment (kirklandsign) [inline examples/arm/run.sh:104]: Still got 
```
CMake Error at CMakeLists.txt:34 (find_package):
  Could not find a package configuration file provided by "executorch" with
  any of the following names:

    executorchConfig.cmake
    executorch-config.cmake

  Add the installation prefix of "executorch" to CMAKE_PREFIX_PATH or set
  "executorch_DIR" to a directory containing one of the above files.  If
  "executorch" provides a separate development package or SDK, be sure it has
  been installed.
```

File: examples/arm/run.sh (hunk 8 above)
Comment (kirklandsign) [reply examples/arm/run.sh]: I got trouble with finding package

File: examples/arm/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+target_include_directories(executorch INTERFACE ${_common_include_directories})
+
+include(${EXECUTORCH_ROOT}/build/Utils.cmake)
+include(${EXECUTORCH_ROOT}/build/Codegen.cmake)
+
+# Generate C++ bindings to register kernels into both PyTorch (for AOT) and
+# Executorch (for runtime). Here select all ops in functions.yaml
+gen_selected_ops("" "${EXECUTORCH_SELECT_OPS_LIST}" "")
+generate_bindings_for_kernels(${EXECUTORCH_ROOT}/kernels/portable/functions.yaml "")
+gen_operators_lib("arm_ops_lib" portable_kernels executorch)
```
Comment (digantdesai) [inline examples/arm/CMakeLists.txt:44]:
```suggestion
gen_operators_lib("portable_ops_lib" portable_kernels executorch)
```
Why add `arm`, seems like something in the op logic which is arm specific?

File: examples/arm/run.sh [hunk 11]
```diff
@@ -5,7 +5,7 @@
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
 
-set -eu
+set -eux
```
Comment (digantdesai) [inline examples/arm/run.sh:8]: add `-x` only for CI, this will make it too verbose for end user running this manually

File: examples/arm/run.sh [hunk 12]
```diff
@@ -58,26 +58,35 @@ function build_executorch() {
...
-    cmake --build . -- -j"$((n - 5))"
+    cmake --build ${et_build_dir} --target install --config Release -- -j"$((n - 5))"
+
+    cmake                                                 \
+        -DCMAKE_INSTALL_PREFIX=cmake-out                  \
+        -DCMAKE_BUILD_TYPE=Release                        \
+        -DEXECUTORCH_SELECT_OPS_LIST="aten::_softmax.out" \
+        -DCMAKE_TOOLCHAIN_FILE="${toolchain_cmake}"       \
+        -B"${et_build_dir}"/examples/arm                  \
+        -Dexecutorch_DIR="${et_root_dir}"/build           \
```
Comment (digantdesai) [inline examples/arm/run.sh:85]:
```suggestion
        -DEXECUTORCH_DIR="${et_root_dir}"/build           \
```
do you mean this?

File: examples/arm/CMakeLists.txt (hunk 10 above)
Comment (kirklandsign) [reply examples/arm/CMakeLists.txt:44]: Nope. It's just a name for op registration lib.  Here it's not registering full portable. It's arm only.

File: examples/arm/CMakeLists.txt (hunk 4 above)
Comment (larryliu0820) [reply examples/arm/CMakeLists.txt]: `executor_runner.cpp` uses a bunch of C++ 17 features
